Unable to execute requests after API token timed out

When using the infocmdb client for a long time and the API token lifespan is over, no more requests with the client will succeed:
```log
[ERROR] Error: github.com/infonova/infocmdb-sdk-go/infocmdb.(*Client).GetAttributeIdByAttributeName: Forbidden
Not authenticated
```

As a workaround the current client could be dropped and a new one can be created, after parsing the error message and matching it against the "Not authenticated" string. This is error prone and unnecessarily complicated.

One possibility is to make the API token lifespan configurable, but this does not solve the root cause and it would still be possible to run into this error (token manually invalidated, ...). It also is not very secure to keep the token alive forever.

Instead the sdk should try to re-login and only return such an error after unsuccessful authentication.

--- a/infocmdb/v2/infocmdb/ci.go
+++ b/infocmdb/v2/infocmdb/ci.go
@@ -7,15 +7,17 @@ import (
 	"strconv"
 )
 
-func (i *Cmdb) CiListByCiTypeID(ciTypeID int, out interface{}) (err error) {
+func (cmdb *Cmdb) CiListByCiTypeID(ciTypeID int, out interface{}) (err error) {
 	var respErr client.ResponseError
-	resp, err := i.Client.NewRequest().
-		SetResult(&out).
-		SetError(&respErr).
-		SetQueryParams(map[string]string{
-			"ciTypeId": fmt.Sprintf("%d", ciTypeID),
-		}).
-		Get("/apiV2/ci/index")
+	resp, err := cmdb.Client.Execute(resty.MethodGet, "/apiV2/ci/index",
+		func(request *resty.Request) *resty.Request {
+			return request.
+				SetResult(&out).
+				SetError(&respErr).
+				SetQueryParams(map[string]string{
+					"ciTypeId": fmt.Sprintf("%d", ciTypeID),
+				})
+		})
 
 	if resp != nil && resp.IsError() {
 		return respErr
@@ -200,19 +202,22 @@ type GetCiDetailResponse struct {
 	} `json:"data"`
 }
 
-func (i *Cmdb) CiDetailByCiId(ciId int64) (ciDetail GetCiDetailResponse, restyRes *resty.Response, err error) {
-	if err = i.Login(); err != nil {
+func (cmdb *Cmdb) CiDetailByCiId(ciId int64) (ciDetail GetCiDetailResponse, restyRes *resty.Response, err error) {
+	if err = cmdb.Login(); err != nil {
 		return
 	}
 
 	var respErr client.ResponseError
 
-	resp, err := i.Client.NewRequest().
-		SetResult(&ciDetail).
-		SetError(&respErr).
-		SetQueryParams(map[string]string{
-			"id": strconv.FormatInt(ciId, 10),
-		}).Get("/apiV2/ci")
+	resp, err := cmdb.Client.Execute(resty.MethodGet, "/apiV2/ci",
+		func(request *resty.Request) *resty.Request {
+			return request.
+				SetResult(&ciDetail).
+				SetError(&respErr).
+				SetQueryParams(map[string]string{
+					"id": strconv.FormatInt(ciId, 10),
+				})
+		})
 
 	if resp != nil && resp.IsError() {
 		return ciDetail, restyRes, respErr
@@ -237,17 +242,18 @@ type updateCiAttributesRequest struct {
 	Ci updateCiAttributes `json:"ci"`
 }
 
-func (i *Cmdb) UpdateCiAttribute(ci int, ua []UpdateCiAttribute) (err error) {
-	if err = i.Login(); err != nil {
+func (cmdb *Cmdb) UpdateCiAttribute(ci int, ua []UpdateCiAttribute) (err error) {
+	if err = cmdb.Login(); err != nil {
 		return
 	}
 
 	var errResp client.ResponseError
-	resp, err := i.Client.NewRequest().
-		SetBody(updateCiAttributesRequest{Ci: updateCiAttributes{Attributes: ua}}).
-		SetAuthToken(i.Config.ApiKey).
-		SetError(&errResp).
-		Put(fmt.Sprintf("/apiV2/ci/%d", ci))
+	resp, err := cmdb.Client.Execute(resty.MethodPut, fmt.Sprintf("/apiV2/ci/%d", ci),
+		func(request *resty.Request) *resty.Request {
+			return request.
+				SetBody(updateCiAttributesRequest{Ci: updateCiAttributes{Attributes: ua}}).
+				SetError(&errResp)
+		})
 
 	if err != nil {
 		return err

--- a/infocmdb/v2/infocmdb/ci_test.go
+++ b/infocmdb/v2/infocmdb/ci_test.go
@@ -157,15 +157,18 @@ func TestInfoCMDB_CiListByCiTypeID(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			c := New()
-			ut.SetValidConfig(&c.Config)
+			cmdb := New()
+			ut.SetValidConfig(&cmdb.Config)
+			if err := cmdb.LoadConfig(cmdb.Config); err != nil {
+				t.Fatalf("LoadConfig failed: %v\n", err)
+			}
 
-			if err := c.Login(); err != nil {
+			if err := cmdb.Login(); err != nil {
 				t.Fatalf("Login failed: %v\n", err)
 			}
 
 			var citypetest2 EmployeeReturn
-			if err := c.CiListByCiTypeID(tt.args.ciTypeID, &citypetest2); (err != nil) != tt.wantErr {
+			if err := cmdb.CiListByCiTypeID(tt.args.ciTypeID, &citypetest2); (err != nil) != tt.wantErr {
 				t.Fatalf("Cmdb.CiListByCiTypeID() error = %v, wantErr %v", err, tt.wantErr)
 			}
 		})

--- a/infocmdb/v2/infocmdb/client/client.go
+++ b/infocmdb/v2/infocmdb/client/client.go
@@ -1,12 +1,23 @@
 package client
 
 import (
+	"errors"
 	"fmt"
+	log "github.com/sirupsen/logrus"
 	"gopkg.in/resty.v1"
+	"strconv"
+	"strings"
 )
 
+type LoginParams struct {
+	Username string
+	Password string
+	Lifetime int
+}
+
 type Client struct {
-	resty *resty.Client
+	resty       *resty.Client
+	loginParams LoginParams
 }
 
 // Response is the default json return of the cmdb upon any request success or error
@@ -21,6 +32,10 @@ type ResponseError struct {
 	Response
 }
 
+func (res ResponseError) Error() string {
+	return fmt.Sprintf("%s\n%+v", res.Message, res.Data)
+}
+
 func New(baseURL string) (c *Client) {
 	c = &Client{}
 	c.resty = resty.New().
@@ -28,14 +43,78 @@ func New(baseURL string) (c *Client) {
 	return
 }
 
-func (c *Client) NewRequest() *resty.Request {
-	return c.resty.NewRequest()
+type loginTokenReturn struct {
+	Success bool   `json:"success"`
+	Message string `json:"message"`
+	Data    struct {
+		Token string `json:"token"`
+	} `json:"data"`
 }
 
-func (c *Client) SetAuthToken(token string) {
+func (c *Client) Login(loginParams LoginParams) (token string, err error) {
+	c.loginParams = loginParams
+
+	if loginParams.Username == "" || loginParams.Password == "" {
+		return "", errors.New("must provide credentials")
+	}
+
+	var loginResult loginTokenReturn
+	params := map[string]string{
+		"username": loginParams.Username,
+		"password": loginParams.Password,
+		"lifetime": strconv.Itoa(loginParams.Lifetime),
+	}
+
+	var errResp ResponseError
+	resp, err := c.resty.
+		NewRequest().
+		SetError(&errResp).
+		SetResult(&loginResult).
+		SetFormData(params).
+		Post("/apiV2/auth/token")
+
+	if err != nil {
+		return "", err
+	}
+
+	if resp != nil && resp.IsError() {
+		return "", errResp
+	}
+
+	token = loginResult.Data.Token
+
+	if token == "" {
+		return "", errors.New("login status not ok")
+	}
+
 	c.resty.SetAuthToken(token)
+	return
 }
 
-func (res ResponseError) Error() string {
-	return fmt.Sprintf("%s\n%+v", res.Message, res.Data)
+type PrepareRequestFunc func(request *resty.Request) *resty.Request
+
+// Executes a request, automatically resolving timed out API token problems and retrying.
+func (c *Client) Execute(method, url string, prepareRequestFunc PrepareRequestFunc) (resp *resty.Response, err error) {
+	req := prepareRequestFunc(c.resty.NewRequest())
+	resp, err = req.Execute(method, url)
+
+	if err != nil {
+		return
+	}
+
+	if resp.IsError() &&
+		resp.StatusCode() == 403 &&
+		strings.Contains(resp.String(), "Not authenticated") {
+
+		log.Debug("Request failed due to authentication error, logging in again and retrying...")
+		token, err := c.Login(c.loginParams)
+		if err != nil {
+			return nil, errors.New("re-login after authentication error failed: " + err.Error())
+		}
+
+		req.SetAuthToken(token)
+		return req.Execute(method, url)
+	}
+
+	return
 }

--- a/infocmdb/v2/infocmdb/cmdb.go
+++ b/infocmdb/v2/infocmdb/cmdb.go
@@ -16,7 +16,6 @@ type Config struct {
 	Url      string `yaml:"apiUrl"`
 	Username string `yaml:"apiUser"`
 	Password string `yaml:"apiPassword"`
-	ApiKey   string
 	BasePath string `yaml:"BasePath"`
 }
 
@@ -70,29 +69,35 @@ const (
 	UPDATE_MODE_SET               = "set"
 )
 
-func (i *Cmdb) LoadConfigFile(path string) (err error) {
-	err = config.LoadYamlConfig(path, &i.Config)
-	if err != nil {
-		return
-	}
-
-	err = i.applyUrlFromRedirect()
-	if err != nil {
-		return
-	}
-
-	log.Debugf("Config after applied url from redirect: %+v", i.Config)
-	i.Client = client.New(i.Config.Url)
+func (cmdb *Cmdb) LoadConfig(config Config) (err error) {
+	cmdb.Config = config
+	cmdb.Client = client.New(config.Url)
 	return
 }
 
-func (i *Cmdb) applyUrlFromRedirect() (err error) {
+func (cmdb *Cmdb) LoadConfigFile(path string) (err error) {
+	err = config.LoadYamlConfig(path, &cmdb.Config)
+	if err != nil {
+		return
+	}
+
+	err = cmdb.applyUrlFromRedirect()
+	if err != nil {
+		return
+	}
+
+	log.Debugf("Config after applied url from redirect: %+v", cmdb.Config)
+	cmdb.Client = client.New(cmdb.Config.Url)
+	return
+}
+
+func (cmdb *Cmdb) applyUrlFromRedirect() (err error) {
 	c := &http.Client{
 		CheckRedirect: func(req *http.Request, via []*http.Request) error {
 			return http.ErrUseLastResponse
 		},
 	}
-	resp, err := c.Get(i.Config.Url)
+	resp, err := c.Get(cmdb.Config.Url)
 	if err != nil {
 		return
 	}
@@ -100,7 +105,7 @@ func (i *Cmdb) applyUrlFromRedirect() (err error) {
 	for name, header := range resp.Header {
 		if name == "Location" && len(header) > 0 {
 			baseUrl, _ := url.Parse(header[0])
-			i.Config.Url = baseUrl.Scheme + "://" + baseUrl.Host + "/"
+			cmdb.Config.Url = baseUrl.Scheme + "://" + baseUrl.Host + "/"
 		}
 	}
 
@@ -108,7 +113,7 @@ func (i *Cmdb) applyUrlFromRedirect() (err error) {
 }
 
 // New returns a new Cmdb Client to access the V2 Api
-func New() (i *Cmdb) {
+func New() (cmdb *Cmdb) {
 	return &Cmdb{
 		Cache:  cache.New(5*time.Minute, 10*time.Minute),
 		Client: &client.Client{},

--- a/infocmdb/v2/infocmdb/login.go
+++ b/infocmdb/v2/infocmdb/login.go
@@ -1,58 +1,29 @@
 package infocmdb
 
 import (
-	"github.com/infonova/infocmdb-sdk-go/infocmdb/v1/infocmdb"
 	"github.com/infonova/infocmdb-sdk-go/infocmdb/v2/infocmdb/client"
+	utilCache "github.com/patrickmn/go-cache"
 	log "github.com/sirupsen/logrus"
 )
 
-type LoginTokenReturn struct {
-	Data struct {
-		Token string `json:"token"`
-	} `json:"data"`
-	Message string `json:"message"`
-	Success bool   `json:"success"`
-}
+func (cmdb *Cmdb) Login() (err error) {
+	cacheKey := "LoggedIn"
+	_, alreadyLoggedIn := cmdb.Cache.Get(cacheKey)
 
-func (i *Cmdb) Login() (err error) {
-
-	if i.Config.ApiKey != "" {
+	if alreadyLoggedIn {
 		log.Trace("already logged in")
 		return nil
 	}
-	if i.Config.Username == "" || i.Config.Password == "" {
-		return infocmdb.ErrNoCredentials
-	}
 
-	var loginResult LoginTokenReturn
-	params := map[string]string{
-		"username": i.Config.Username,
-		"password": i.Config.Password,
-		"lifetime": "600",
-	}
-
-	i.Client = client.New(i.Config.Url)
-
-	var errResp client.ResponseError
-	resp, err := i.Client.NewRequest().
-		SetError(&errResp).
-		SetResult(&loginResult).
-		SetFormData(params).
-		Post("/apiV2/auth/token")
-
+	_, err = cmdb.Client.Login(client.LoginParams{
+		Username: cmdb.Config.Username,
+		Password: cmdb.Config.Password,
+		Lifetime: 600,
+	})
 	if err != nil {
-		return err
+		return
 	}
 
-	if resp != nil && resp.IsError() {
-		return errResp
-	}
-
-	if loginResult.Data.Token == "" {
-		return infocmdb.ErrLoginFailed
-	}
-
-	i.Config.ApiKey = loginResult.Data.Token
-	i.Client.SetAuthToken(i.Config.ApiKey)
+	cmdb.Cache.Set(cacheKey, true, utilCache.NoExpiration)
 	return
 }

--- a/infocmdb/v2/infocmdb/login_test.go
+++ b/infocmdb/v2/infocmdb/login_test.go
@@ -3,9 +3,7 @@ package infocmdb
 import (
 	"testing"
 
-	"github.com/infonova/infocmdb-sdk-go/infocmdb/v2/infocmdb/client"
 	utilTesting "github.com/infonova/infocmdb-sdk-go/util/testing"
-	"github.com/patrickmn/go-cache"
 )
 
 func TestInfoCMDB_LoginWithUserPass(t *testing.T) {
@@ -13,8 +11,6 @@ func TestInfoCMDB_LoginWithUserPass(t *testing.T) {
 
 	type fields struct {
 		Config Config
-		Cache  *cache.Cache
-		Client *client.Client
 	}
 	type args struct {
 		url      string
@@ -29,19 +25,19 @@ func TestInfoCMDB_LoginWithUserPass(t *testing.T) {
 	}{
 		{
 			"valid login admin//admin",
-			fields{Config: Config{Url: url, Username: "admin", Password: "admin"}},
+			fields{Config{Url: url, Username: "admin", Password: "admin"}},
 			args{},
 			false,
 		},
 		{
 			"invalid login admin//noadmin",
-			fields{Config: Config{Url: url, Username: "admin", Password: "fail"}},
+			fields{Config{Url: url, Username: "admin", Password: "fail"}},
 			args{},
 			true,
 		},
 		{
 			"invalid no username",
-			fields{Config: Config{Url: url, Username: "", Password: "fail"}},
+			fields{Config{Url: url, Username: "", Password: "fail"}},
 			args{},
 			true,
 		},
@@ -60,12 +56,12 @@ func TestInfoCMDB_LoginWithUserPass(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			i := new(Cmdb)
-			i.Config = tt.fields.Config
-			i.Cache = tt.fields.Cache
-			i.Client = client.New(i.Config.Url)
+			cmdbV2 := New()
+			if err := cmdbV2.LoadConfig(tt.fields.Config); err != nil {
+				t.Fatalf("LoadConfig failed: %v\n", err)
+			}
 
-			if err := i.Login(); (err != nil) != tt.wantErr {
+			if err := cmdbV2.Login(); (err != nil) != tt.wantErr {
 				t.Errorf("Cmdb.LoginWithUserPass() error = %v, wantErr %v", err, tt.wantErr)
 			}
 		})

--- a/infocmdb/v2/infocmdb/query.go
+++ b/infocmdb/v2/infocmdb/query.go
@@ -3,6 +3,7 @@ package infocmdb
 import (
 	"github.com/infonova/infocmdb-sdk-go/infocmdb/v2/infocmdb/client"
 	log "github.com/sirupsen/logrus"
+	"gopkg.in/resty.v1"
 )
 
 type queryParams struct {
@@ -13,14 +14,14 @@ type queryRequest struct {
 	Query queryParams `json:"query"`
 }
 
-func (i *Cmdb) Query(query string, out interface{}, params map[string]string) (err error) {
+func (cmdb *Cmdb) Query(query string, out interface{}, params map[string]string) (err error) {
 	log.Debugf("Querying webservice %v with params %v", query, params)
 
-	if err = i.Login(); err != nil {
+	if err = cmdb.Login(); err != nil {
 		return
 	}
 
-	r := queryRequest{
+	body := queryRequest{
 		Query: queryParams{
 			Params: params,
 		},
@@ -28,19 +29,20 @@ func (i *Cmdb) Query(query string, out interface{}, params map[string]string) (e
 
 	var respError client.ResponseError
 
-	resp, err := i.Client.NewRequest().
-		SetResult(out).
-		SetBody(r).
-		SetAuthToken(i.Config.ApiKey).
-		SetError(&respError).
-		Put("/apiV2/query/execute/" + query)
+	resp, err := cmdb.Client.Execute(resty.MethodPut, "/apiV2/query/execute/"+query,
+		func(request *resty.Request) *resty.Request {
+			return request.
+				SetResult(out).
+				SetBody(body).
+				SetError(&respError)
+		})
 
 	if err != nil {
 		return
 	}
 
 	if resp.IsError() {
-		log.Debugf("Error result: %v", respError)
+		log.Debugf("Status: %v, Error result: %v", resp.StatusCode(), respError)
 		return respError
 	}
 
@@ -49,8 +51,8 @@ func (i *Cmdb) Query(query string, out interface{}, params map[string]string) (e
 	return
 }
 
-func (i *Cmdb) QueryRaw(query string, params map[string]string) (r string, err error) {
-	if err = i.Login(); err != nil {
+func (cmdb *Cmdb) QueryRaw(query string, params map[string]string) (r string, err error) {
+	if err = cmdb.Login(); err != nil {
 		return
 	}
 
@@ -61,11 +63,12 @@ func (i *Cmdb) QueryRaw(query string, params map[string]string) (r string, err e
 	}
 
 	var respError client.ResponseError
-	resp, err := i.Client.NewRequest().
-		SetBody(qr).
-		SetAuthToken(i.Config.ApiKey).
-		SetError(&respError).
-		Put("/apiV2/query/execute/" + query)
+	resp, err := cmdb.Client.Execute(resty.MethodPut, "/apiV2/query/execute/"+query,
+		func(request *resty.Request) *resty.Request {
+			return request.
+				SetBody(qr).
+				SetError(&respError)
+		})
 
 	if resp == nil {
 		return

--- a/infocmdb/v2/infocmdb/query_test.go
+++ b/infocmdb/v2/infocmdb/query_test.go
@@ -3,9 +3,7 @@ package infocmdb
 import (
 	"testing"
 
-	"github.com/infonova/infocmdb-sdk-go/infocmdb/v2/infocmdb/client"
 	utilTesting "github.com/infonova/infocmdb-sdk-go/util/testing"
-	"github.com/patrickmn/go-cache"
 )
 
 func TestInfoCMDB_Query(t *testing.T) {
@@ -13,9 +11,6 @@ func TestInfoCMDB_Query(t *testing.T) {
 
 	type fields struct {
 		Config Config
-		Cache  *cache.Cache
-		Client *client.Client
-		Error  error
 	}
 	type args struct {
 		query  string
@@ -38,17 +33,16 @@ func TestInfoCMDB_Query(t *testing.T) {
 		args    args
 		wantErr bool
 	}{
-		{"valid", fields{Config: Config{Url: url, Username: "admin", Password: "admin"}}, a, false},
+		{"valid", fields{Config{Url: url, Username: "admin", Password: "admin"}}, a, false},
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			i := &Cmdb{
-				Config: tt.fields.Config,
-				Cache:  tt.fields.Cache,
-				Client: client.New(tt.fields.Config.Url),
-				Error:  tt.fields.Error,
+			cmdbV2 := New()
+			if err := cmdbV2.LoadConfig(tt.fields.Config); err != nil {
+				t.Fatalf("LoadConfig failed: %v\n", err)
 			}
-			if err := i.Query(tt.args.query, tt.args.out, tt.args.params); (err != nil) != tt.wantErr {
+
+			if err := cmdbV2.Query(tt.args.query, tt.args.out, tt.args.params); (err != nil) != tt.wantErr {
 				t.Errorf("Cmdb.Query() error = %v, wantErr %v", err, tt.wantErr)
 			}
 		})

--- a/infocmdb/v2/infocmdb/workflow_context.go
+++ b/infocmdb/v2/infocmdb/workflow_context.go
@@ -160,8 +160,8 @@ type getWorkflowContextResponse struct {
 	} `json:"data"`
 }
 
-func (i *Cmdb) GetWorkflowContext(workflowInstanceId int) (workflowContext *WorkflowContext, err error) {
-	if err = i.Login(); err != nil {
+func (cmdb *Cmdb) GetWorkflowContext(workflowInstanceId int) (workflowContext *WorkflowContext, err error) {
+	if err = cmdb.Login(); err != nil {
 		return
 	}
 
@@ -170,7 +170,7 @@ func (i *Cmdb) GetWorkflowContext(workflowInstanceId int) (workflowContext *Work
 	}
 
 	getWorkflowContextResponse := getWorkflowContextResponse{}
-	err = i.Query("int_getWorkflowContext", &getWorkflowContextResponse, params)
+	err = cmdb.Query("int_getWorkflowContext", &getWorkflowContextResponse, params)
 	if err != nil {
 		return
 	}

--- a/infocmdb/webservice_default_test.go
+++ b/infocmdb/webservice_default_test.go
@@ -1,22 +1,16 @@
 package infocmdb
 
 import (
-	"github.com/patrickmn/go-cache"
+	v2 "github.com/infonova/infocmdb-sdk-go/infocmdb/v2/infocmdb"
+	utilTesting "github.com/infonova/infocmdb-sdk-go/util/testing"
 	"reflect"
 	"testing"
-	"time"
-
-	v1 "github.com/infonova/infocmdb-sdk-go/infocmdb/v1/infocmdb"
-	v2 "github.com/infonova/infocmdb-sdk-go/infocmdb/v2/infocmdb"
-
-	utilTesting "github.com/infonova/infocmdb-sdk-go/util/testing"
 )
 
 func TestInfoCMDB_CreateCi(t *testing.T) {
 	infocmdbUrl := utilTesting.New().GetUrl()
 	type fields struct {
-		v1 *v1.Cmdb
-		v2 *v2.Cmdb
+		v2Config v2.Config
 	}
 	type args struct {
 		ciTypeID  int
@@ -33,13 +27,12 @@ func TestInfoCMDB_CreateCi(t *testing.T) {
 		{
 			"v2 create CI",
 			fields{
-				&v1.Cmdb{Config: v1.Config{}},
-				&v2.Cmdb{Config: v2.Config{
+				v2.Config{
 					Url:      infocmdbUrl,
 					Username: "admin",
 					Password: "admin",
 					BasePath: "/app/",
-				}},
+				},
 			},
 			args{
 				ciTypeID:  476,
@@ -60,11 +53,14 @@ func TestInfoCMDB_CreateCi(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			i := &Client{
-				v1: tt.fields.v1,
-				v2: tt.fields.v2,
+			cmdbV2 := v2.New()
+			if err := cmdbV2.LoadConfig(tt.fields.v2Config); err != nil {
+				t.Fatalf("LoadConfig failed: %v\n", err)
 			}
-			gotR, err := i.CreateCi(tt.args.ciTypeID, tt.args.icon, tt.args.historyId)
+			cmdb := &Client{
+				v2: cmdbV2,
+			}
+			gotR, err := cmdb.CreateCi(tt.args.ciTypeID, tt.args.icon, tt.args.historyId)
 			if (err != nil) != tt.wantErr {
 				t.Errorf("CreateCi() error = %v, wantErr %v", err, tt.wantErr)
 				return
@@ -79,8 +75,7 @@ func TestInfoCMDB_CreateCi(t *testing.T) {
 func TestInfoCMDB_AddCiProjectMapping(t *testing.T) {
 	infocmdbUrl := utilTesting.New().GetUrl()
 	type fields struct {
-		v1 *v1.Cmdb
-		v2 *v2.Cmdb
+		v2Config v2.Config
 	}
 	type args struct {
 		ciID      int
@@ -96,13 +91,12 @@ func TestInfoCMDB_AddCiProjectMapping(t *testing.T) {
 		{
 			"v2 create CI",
 			fields{
-				&v1.Cmdb{Config: v1.Config{}},
-				&v2.Cmdb{Config: v2.Config{
+				v2.Config{
 					Url:      infocmdbUrl,
 					Username: "admin",
 					Password: "admin",
 					BasePath: "/app/",
-				}},
+				},
 			},
 			args{
 				ciID:      617830,
@@ -114,11 +108,14 @@ func TestInfoCMDB_AddCiProjectMapping(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			i := &Client{
-				v1: tt.fields.v1,
-				v2: tt.fields.v2,
+			cmdbV2 := v2.New()
+			if err := cmdbV2.LoadConfig(tt.fields.v2Config); err != nil {
+				t.Fatalf("LoadConfig failed: %v\n", err)
 			}
-			err := i.AddCiProjectMapping(tt.args.ciID, tt.args.projectID, tt.args.historyID)
+			cmdb := &Client{
+				v2: cmdbV2,
+			}
+			err := cmdb.AddCiProjectMapping(tt.args.ciID, tt.args.projectID, tt.args.historyID)
 			if (err != nil) != tt.wantErr {
 				t.Errorf("AddCiProjectMapping() error = %v, wantErr %v", err, tt.wantErr)
 				return
@@ -130,8 +127,7 @@ func TestInfoCMDB_AddCiProjectMapping(t *testing.T) {
 func TestInfoCMDB_GetListOfCiIdsOfCiType(t *testing.T) {
 	infocmdbUrl := utilTesting.New().GetUrl()
 	type fields struct {
-		v1 *v1.Cmdb
-		v2 *v2.Cmdb
+		v2Config v2.Config
 	}
 	type args struct {
 		ciTypeID int
@@ -146,13 +142,12 @@ func TestInfoCMDB_GetListOfCiIdsOfCiType(t *testing.T) {
 		{
 			"v2 List Ci's pf Type '1' with wrong Credentials (fail)",
 			fields{
-				&v1.Cmdb{Config: v1.Config{}},
-				&v2.Cmdb{Config: v2.Config{
+				v2.Config{
 					Url:      infocmdbUrl,
 					Username: "fail",
 					Password: "fail",
 					BasePath: "/app/",
-				}},
+				},
 			},
 			args{ciTypeID: 1},
 			nil,
@@ -161,13 +156,12 @@ func TestInfoCMDB_GetListOfCiIdsOfCiType(t *testing.T) {
 		{
 			"v2 List Ci's of Type '1' (demo)",
 			fields{
-				&v1.Cmdb{Config: v1.Config{}},
-				&v2.Cmdb{Config: v2.Config{
+				v2.Config{
 					Url:      infocmdbUrl,
 					Username: "admin",
 					Password: "admin",
 					BasePath: "/app/",
-				}},
+				},
 			},
 			args{ciTypeID: 1},
 			CiIds{1, 2},
@@ -176,13 +170,12 @@ func TestInfoCMDB_GetListOfCiIdsOfCiType(t *testing.T) {
 		{
 			"v2 List Ci's of Type '-1' (error)",
 			fields{
-				&v1.Cmdb{Config: v1.Config{}},
-				&v2.Cmdb{Config: v2.Config{
+				v2.Config{
 					Url:      infocmdbUrl,
 					Username: "admin",
 					Password: "admin",
 					BasePath: "/app/",
-				}},
+				},
 			},
 			args{ciTypeID: -1},
 			nil,
@@ -191,11 +184,14 @@ func TestInfoCMDB_GetListOfCiIdsOfCiType(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			i := &Client{
-				v1: tt.fields.v1,
-				v2: tt.fields.v2,
+			cmdbV2 := v2.New()
+			if err := cmdbV2.LoadConfig(tt.fields.v2Config); err != nil {
+				t.Fatalf("LoadConfig failed: %v\n", err)
 			}
-			gotR, err := i.GetListOfCiIdsOfCiType(tt.args.ciTypeID)
+			cmdb := &Client{
+				v2: cmdbV2,
+			}
+			gotR, err := cmdb.GetListOfCiIdsOfCiType(tt.args.ciTypeID)
 			if (err != nil) != tt.wantErr {
 				t.Errorf("getListOfCiIdsOfCiType() error = %v, wantErr %v", err, tt.wantErr)
 				return
@@ -210,8 +206,7 @@ func TestInfoCMDB_GetListOfCiIdsOfCiType(t *testing.T) {
 func TestInfoCMDB_GetListOfCiIdsOfCiTypeV2(t *testing.T) {
 	infocmdbUrl := utilTesting.New().GetUrl()
 	type fields struct {
-		v1 *v1.Cmdb
-		v2 *v2.Cmdb
+		v2Config v2.Config
 	}
 	type args struct {
 		ciTypeID int
@@ -226,13 +221,12 @@ func TestInfoCMDB_GetListOfCiIdsOfCiTypeV2(t *testing.T) {
 		{
 			"v2 List Ci's pf Type '1' with wrong Credentials (fail)",
 			fields{
-				&v1.Cmdb{Config: v1.Config{}},
-				&v2.Cmdb{Config: v2.Config{
+				v2.Config{
 					Url:      infocmdbUrl,
 					Username: "false",
 					Password: "false",
 					BasePath: "/app/",
-				}},
+				},
 			},
 			args{ciTypeID: 1},
 			nil,
@@ -241,13 +235,12 @@ func TestInfoCMDB_GetListOfCiIdsOfCiTypeV2(t *testing.T) {
 		{
 			"v2 List CIs of Type 1 (demo)",
 			fields{
-				&v1.Cmdb{Config: v1.Config{}},
-				&v2.Cmdb{Config: v2.Config{
+				v2.Config{
 					Url:      infocmdbUrl,
 					Username: "admin",
 					Password: "admin",
 					BasePath: "/app/",
-				}},
+				},
 			},
 			args{ciTypeID: 1},
 			CiIds{1, 2},
@@ -256,13 +249,12 @@ func TestInfoCMDB_GetListOfCiIdsOfCiTypeV2(t *testing.T) {
 		{
 			"v2 List Ci's of Type '-1' (error)",
 			fields{
-				&v1.Cmdb{Config: v1.Config{}},
-				&v2.Cmdb{Config: v2.Config{
+				v2.Config{
 					Url:      infocmdbUrl,
 					Username: "admin",
 					Password: "admin",
 					BasePath: "/app/",
-				}},
+				},
 			},
 			args{ciTypeID: -1},
 			nil,
@@ -271,11 +263,14 @@ func TestInfoCMDB_GetListOfCiIdsOfCiTypeV2(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			i := &Client{
-				v1: tt.fields.v1,
-				v2: tt.fields.v2,
+			cmdbV2 := v2.New()
+			if err := cmdbV2.LoadConfig(tt.fields.v2Config); err != nil {
+				t.Fatalf("LoadConfig failed: %v\n", err)
 			}
-			gotR, err := i.GetListOfCiIdsOfCiTypeV2(tt.args.ciTypeID)
+			cmdb := &Client{
+				v2: cmdbV2,
+			}
+			gotR, err := cmdb.GetListOfCiIdsOfCiTypeV2(tt.args.ciTypeID)
 			if (err != nil) != tt.wantErr {
 				t.Errorf("getListOfCiIdsOfCiType() error = %v, wantErr %v", err, tt.wantErr)
 				return
@@ -290,7 +285,7 @@ func TestInfoCMDB_GetListOfCiIdsOfCiTypeV2(t *testing.T) {
 func TestInfoCMDB_QueryWebservice(t *testing.T) {
 	infocmdbUrl := utilTesting.New().GetUrl()
 	type fields struct {
-		v2 *v2.Cmdb
+		v2Config v2.Config
 	}
 	type args struct {
 		ws     string
@@ -307,12 +302,12 @@ func TestInfoCMDB_QueryWebservice(t *testing.T) {
 		{
 			"v2 List CIs of Type 1 (demo)",
 			fields{
-				&v2.Cmdb{Config: v2.Config{
+				v2.Config{
 					Url:      infocmdbUrl,
 					Username: "admin",
 					Password: "admin",
 					BasePath: "/app/",
-				}},
+				},
 			},
 			args{ws: "int_getCi", params: map[string]string{"argv1": "1"}},
 			`{"success":true,"message":"Query executed successfully","data":[{"ci_id":"1","ci_type_id":"1","ci_type":"demo","project":"springfield","project_id":"4"}]}`,
@@ -321,10 +316,14 @@ func TestInfoCMDB_QueryWebservice(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			i := &Client{
-				v2: tt.fields.v2,
+			cmdbV2 := v2.New()
+			if err := cmdbV2.LoadConfig(tt.fields.v2Config); err != nil {
+				t.Fatalf("LoadConfig failed: %v\n", err)
 			}
-			gotR, err := i.QueryWebservice(tt.args.ws, tt.args.params)
+			cmdb := &Client{
+				v2: cmdbV2,
+			}
+			gotR, err := cmdb.QueryWebservice(tt.args.ws, tt.args.params)
 			if (err != nil) != tt.wantErr {
 				t.Errorf("QueryWebservice() error = %v, wantErr %v", err, tt.wantErr)
 				return
@@ -339,19 +338,19 @@ func TestInfoCMDB_QueryWebservice(t *testing.T) {
 func TestInfoCMDB_UpdateCiAttribute(t *testing.T) {
 	infocmdbUrl := utilTesting.New().GetUrl()
 	type fields struct {
-		v2 *v2.Cmdb
+		v2Config v2.Config
 	}
 	type args struct {
 		ci int
 		ua []v2.UpdateCiAttribute
 	}
 
-	cmdbConfigValid := v2.Cmdb{Config: v2.Config{
+	cmdbConfigValid := v2.Config{
 		Url:      infocmdbUrl,
 		Username: "admin",
 		Password: "admin",
 		BasePath: "/app/",
-	}}
+	}
 	baseCiID := 14 // ci to base this tests on
 	baseAttributeName := "emp_lastname"
 
@@ -363,7 +362,7 @@ func TestInfoCMDB_UpdateCiAttribute(t *testing.T) {
 	}{
 		{
 			"v2 Delete CI Attribute - fail - requires ciattributeid",
-			fields{&cmdbConfigValid},
+			fields{cmdbConfigValid},
 			args{ci: baseCiID, ua: []v2.UpdateCiAttribute{
 				{Mode: v2.UPDATE_MODE_DELETE, Name: baseAttributeName},
 			}},
@@ -371,7 +370,7 @@ func TestInfoCMDB_UpdateCiAttribute(t *testing.T) {
 		},
 		{
 			"v2 Update CI Attribute",
-			fields{&cmdbConfigValid},
+			fields{cmdbConfigValid},
 			args{ci: baseCiID, ua: []v2.UpdateCiAttribute{
 				{Mode: v2.UPDATE_MODE_SET, Name: "emp_firstname", Value: "22322"},
 			}},
@@ -379,7 +378,7 @@ func TestInfoCMDB_UpdateCiAttribute(t *testing.T) {
 		},
 		{
 			"v2 Update CI Attribute - wrong attribute name",
-			fields{&cmdbConfigValid},
+			fields{cmdbConfigValid},
 			args{ci: baseCiID, ua: []v2.UpdateCiAttribute{
 				{Mode: v2.UPDATE_MODE_SET, Name: baseAttributeName + "_NOT_EXISTING", Value: "1"},
 			}},
@@ -387,7 +386,7 @@ func TestInfoCMDB_UpdateCiAttribute(t *testing.T) {
 		},
 		{
 			"v2 Insert CI Attribute",
-			fields{&cmdbConfigValid},
+			fields{cmdbConfigValid},
 			args{ci: baseCiID, ua: []v2.UpdateCiAttribute{
 				{Mode: v2.UPDATE_MODE_INSERT, Name: baseAttributeName, Value: "New1"},
 			}},
@@ -395,7 +394,7 @@ func TestInfoCMDB_UpdateCiAttribute(t *testing.T) {
 		},
 		{
 			"v2 Update CI Attribute - fail - multiple attributes with the same name",
-			fields{&cmdbConfigValid},
+			fields{cmdbConfigValid},
 			args{ci: baseCiID, ua: []v2.UpdateCiAttribute{
 				{Mode: v2.UPDATE_MODE_SET, Name: baseAttributeName, Value: "22322"},
 			}},
@@ -404,10 +403,14 @@ func TestInfoCMDB_UpdateCiAttribute(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			i := &Client{
-				v2: tt.fields.v2,
+			cmdbV2 := v2.New()
+			if err := cmdbV2.LoadConfig(tt.fields.v2Config); err != nil {
+				t.Fatalf("LoadConfig failed: %v\n", err)
 			}
-			err := i.UpdateCiAttribute(tt.args.ci, tt.args.ua)
+			cmdb := &Client{
+				v2: cmdbV2,
+			}
+			err := cmdb.UpdateCiAttribute(tt.args.ci, tt.args.ua)
 			if (err != nil) != tt.wantErr {
 				t.Errorf("UpdateCiAttribute() error = %v, wantErr %v", err, tt.wantErr)
 				return
@@ -419,8 +422,7 @@ func TestInfoCMDB_UpdateCiAttribute(t *testing.T) {
 func TestInfoCMDB_GetAttrDefaultOptionIdByAttrId(t *testing.T) {
 	infocmdbUrl := utilTesting.New().GetUrl()
 	type fields struct {
-		v1 *v1.Cmdb
-		v2 *v2.Cmdb
+		v2Config v2.Config
 	}
 	type args struct {
 		attrId      int
@@ -436,15 +438,11 @@ func TestInfoCMDB_GetAttrDefaultOptionIdByAttrId(t *testing.T) {
 		{
 			"v2 GetAttributeDefaultOptionId",
 			fields{
-				&v1.Cmdb{Config: v1.Config{}},
-				&v2.Cmdb{
-					Config: v2.Config{
-						Url:      infocmdbUrl,
-						Username: "admin",
-						Password: "admin",
-						BasePath: "/app/",
-					},
-					Cache: cache.New(5*time.Minute, 10*time.Minute),
+				v2.Config{
+					Url:      infocmdbUrl,
+					Username: "admin",
+					Password: "admin",
+					BasePath: "/app/",
 				},
 			},
 			args{
@@ -457,11 +455,14 @@ func TestInfoCMDB_GetAttrDefaultOptionIdByAttrId(t *testing.T) {
 	}
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			i := &Client{
-				v1: tt.fields.v1,
-				v2: tt.fields.v2,
+			cmdbV2 := v2.New()
+			if err := cmdbV2.LoadConfig(tt.fields.v2Config); err != nil {
+				t.Fatalf("LoadConfig failed: %v\n", err)
 			}
-			gotR, err := i.GetAttrDefaultOptionIdByAttrId(tt.args.attrId, tt.args.optionValue)
+			cmdb := &Client{
+				v2: cmdbV2,
+			}
+			gotR, err := cmdb.GetAttrDefaultOptionIdByAttrId(tt.args.attrId, tt.args.optionValue)
 			if (err != nil) != tt.wantErr {
 				t.Errorf("GetAttrDefaultOptionIdByAttrId() error = %v, wantErr %v", err, tt.wantErr)
 				return

